<a href="https://colab.research.google.com/github/tronicey/Predictions-of-Product-Sales/blob/main/predictions_of_product_sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predictions of Product Sales
- Author: Tronice Ray

## Project Overview

## Load and Inspect Data

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from google.colab import drive

In [2]:
#Save the file path and load the data as df_sales
fpath = ('/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv')
df_sales = pd.read_csv(fpath)
# Display the first 5 rows
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [3]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


## Clean Data

1. How many rows and columns?

In [4]:
df_sales.shape

(8523, 12)

- There are 8523 rows and 12 columns

2. What are the data types of each variable?

In [5]:
df_sales.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

- Information looks correct.

3. Are there any duplicates? If so, drop any duplicates.

In [6]:
# Check for duplicates
df_sales.duplicated().sum()

0

- There are 0 duplicates

4. Identify missing values.

In [7]:
df_sales.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

- Missing Values
  - Item_Weight is missing 1463 values.
  - Outlet_Size is missing 2410 values.

5. Address the missing values by using a placeholder value.

In [8]:
# save list of categorical column name.
# Saving list of column names for numeric cols
num_cols = df_sales.select_dtypes('number').columns
# use the list of num columns to fill null values with -1
df_sales[num_cols]  = df_sales[num_cols].fillna(-1)
df_sales[num_cols].isna().sum()

Item_Weight                  0
Item_Visibility              0
Item_MRP                     0
Outlet_Establishment_Year    0
Item_Outlet_Sales            0
dtype: int64

- There are no longer any missing values in Item_Weight

In [9]:
# Saving list of column names for object cols
cat_cols = df_sales.select_dtypes('object').columns
# use the list of cat columns to fill null values with MISSING
# Using the list of cat_cols to fillna
df_sales[cat_cols] = df_sales[cat_cols].fillna("MISSING")


6. Confirm there are no missing values after addressing them

In [10]:
# Check for remaining missing values
df_sales.isna().sum()


Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

- After addressing missing sales, there are no missing sales in the data frame.

7. Find and fix any inconsistent categories of data

In [11]:
# Remake the list of string columns
object_cols = df_sales.select_dtypes("object").columns
# loop through the list of string columns
for col in object_cols:
  # print the value counts for the column
  # Explore the value counts of one feature
   print(f"Value Counts for {col}")
   print(df_sales[col].value_counts())
  # Increasing readability by adding an empty line
print('\n')



Value Counts for Item_Identifier
FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64
Value Counts for Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64
Value Counts for Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64
Value Counts for Outlet_Identifier
OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035

- Inconsistencies were found in Item_Fat_Content
  - LF and low fat need to be added to Low Fat
  - Reg needs to be added to Regular

In [12]:
# Replace "LF" with "Low Fat"
df_sales['Item_Fat_Content'] = df_sales['Item_Fat_Content'].replace({'LF':"Low Fat"})
# Replace "low fat" with "Low Fat"
df_sales['Item_Fat_Content'] = df_sales['Item_Fat_Content'].replace({'low fat':"Low Fat"})
# Check for inconsistencies by inspecting the value_counts for Item_Fat_Content
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    2889
reg         117
Name: Item_Fat_Content, dtype: int64

In [13]:
# Replace "reg" with "Regular"
df_sales['Item_Fat_Content'] = df_sales['Item_Fat_Content'].replace({'reg':"Regular"})
# Check for inconsistencies by inspecting the value_counts for Item_Fat_Content
df_sales['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

- All inconsistencies have been removed from data frame

8. For any numerical columns, obtain the summary statistics of each (min, max, mean).

In [14]:
# Pull all numerical columns and round them at two decimal points.
df_sales.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      8523.00          8523.00   8523.00                    8523.00   
mean         10.48             0.07    140.99                    1997.83   
std           6.72             0.05     62.28                       8.37   
min          -1.00             0.00     31.29                    1985.00   
25%           6.65             0.03     93.83                    1987.00   
50%          11.00             0.05    143.01                    1999.00   
75%          16.00             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96

In [16]:
#Find the mean, smallest number and largest number for all number columns and round them to two decimal points.
stats =  df_sales.describe()
stats.loc[['mean','min','max']].round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
mean        10.48             0.07    140.99                    1997.83   
min         -1.00             0.00     31.29                    1985.00   
max         21.35             0.33    266.89                    2009.00   

      Item_Outlet_Sales  
mean            2181.29  
min               33.29  
max            13086.96

- The mean, min and max numbers were found for all number columns.
  - Item weight shows -1 for it's smallest number. It is assumed that is a filler for a missing number.
  - All other columns do not seem to have any issues.

## Exploratory Data Analysis

## Feature Inspection